### ISSUE TITLE
Vastly different results on local Mac OS X vs Linux VM

### ISSUE THREAD

**Issue by jmaddalena** [dmlc/xgboost#3046]
TL;DR -- xgboost generates significantly worse predictions when ran on Mac OS X vs a Linux VM, even when a seed is set, for any values of `subsample` or `colsample_bytree` that are less than 1.0. 

I am experiencing an alarming issue with xgboost that generates vastly different predictions based on whether I'm running the code on my local Mac or on a Linux VM. From [this issue](https://github.com/dmlc/xgboost/issues/310) I was able to isolate the problem to the hyperparameters involving randomness: `subsample` and `colsample_bytree`. When both of these are specified as 1.0 and a seed is set, I obtain the same exact results locally and on my VM. However, varying either one while holding the other constant, even with a set seed, creates results between the operating systems that are too vastly different to be explained by differences in random number generators. 

In particular, my local machine consistently generates much worse predictions and has differing lengths of the raw vector data stored in the `xgb.train()` output model object. Additionally and perhaps most alarmingly, the number of features in the output of `xgb.importance()` is directly proportional to the value specified by `colsample_bytree`. Conversely, all features are included in the importance data frame when ran on the Linux VM, regardless of the specified `colsample_bytree`. My concern with my local OS is that for, say, a training data set of 10 features with a specified `colsample_bytree` of 0.5, only 5 features are utilized *in total*, not just per tree, even if all 10 features are predictive.  

One final oddity: with the data that I originally discovered this issue, I was able to obtain identical results with a set seed once all missing data was removed for varying choices of `subsample` and `colsample_bytree` fixed at 1.0 , making me think `xgboost` handled missingness differently across the two systems. However, the data in my reproducible example has no missingness and I do not obtain identical results for varying levels of `subsample`. No solutions for `colsample_bytree` have yet been uncovered besides setting a seed and leaving this value at 1.0. 

Below you will find my sessionInfo() from both systems and a reproducible example with code, and attached source data and plots of my results of running identical code with set seeds across the two different operating systems. 

**Local OS X**

```> sessionInfo()
R version 3.4.3 (2017-11-30)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS Sierra 10.12.6

Matrix products: default
BLAS: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] bindrcpp_0.2     xgboost_0.6-4    dplyr_0.7.4.9000 purrr_0.2.4      readr_1.1.1      tidyr_0.7.2      tibble_1.3.4    
[8] ggplot2_2.2.1    tidyverse_1.1.1 

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.14      cellranger_1.1.0  compiler_3.4.3    plyr_1.8.4        bindr_0.1         forcats_0.2.0    
 [7] base64enc_0.1-3   tools_3.4.3       digest_0.6.12     lubridate_1.7.1   gtable_0.2.0      jsonlite_1.5     
[13] evaluate_0.10.1   nlme_3.1-131      lattice_0.20-35   pkgconfig_2.0.1   rlang_0.1.6       Matrix_1.2-12    
[19] psych_1.7.5       yaml_2.1.14       parallel_3.4.3    haven_1.0.0       xml2_1.2.0        httr_1.3.1       
[25] stringr_1.2.0     knitr_1.16        hms_0.3           rprojroot_1.2     grid_3.4.3        tidyselect_0.2.3 
[31] glue_1.2.0        data.table_1.10.4 R6_2.2.2          readxl_1.0.0      foreign_0.8-69    rmarkdown_1.6    
[37] modelr_0.1.0      reshape2_1.4.2    magrittr_1.5      scales_0.5.0      backports_1.1.0   htmltools_0.3.6  
[43] rvest_0.3.2       assertthat_0.2.0  mnormt_1.5-5      colorspace_1.3-2  labeling_0.3      stringi_1.1.5    
[49] lazyeval_0.2.0    munsell_0.4.3     broom_0.4.2      
```

**VM (Linux)**

```
> sessionInfo()
R version 3.4.2 (2017-09-28)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 16.04.2 LTS

Matrix products: default
BLAS: /usr/lib/libblas/libblas.so.3.6.0
LAPACK: /usr/lib/lapack/liblapack.so.3.6.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] xgboost_0.6-4   dplyr_0.7.4     purrr_0.2.4     readr_1.1.1    
[5] tidyr_0.7.2     tibble_1.4.1    ggplot2_2.2.1   tidyverse_1.1.1

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.13        cellranger_1.1.0    pillar_1.0.1       
 [4] compiler_3.4.2      plyr_1.8.4          bindr_0.1          
 [7] forcats_0.2.0       tools_3.4.2         lubridate_1.7.1    
[10] jsonlite_1.5        nlme_3.1-131        gtable_0.2.0       
[13] lattice_0.20-35     pkgconfig_2.0.1     rlang_0.1.6        
[16] Matrix_1.2-11       psych_1.7.8         parallel_3.4.2     
[19] haven_1.1.0         bindrcpp_0.2        xml2_1.1.1         
[22] stringr_1.2.0       httr_1.3.1          hms_0.3            
[25] grid_3.4.2          data.table_1.10.4-2 glue_1.1.1         
[28] R6_2.2.2            readxl_1.0.0        foreign_0.8-69     
[31] modelr_0.1.1        reshape2_1.4.2      magrittr_1.5       
[34] scales_0.5.0        rvest_0.3.2         assertthat_0.2.0   
[37] mnormt_1.5-5        colorspace_1.3-2    stringi_1.1.5      
[40] lazyeval_0.2.0      munsell_0.4.3       broom_0.4.2 
```

**Reproducible Example**

For my reproducible example I am using data to predict the daily numbers of burglaries in Chicago, consisting of of only a few time-related features -- month, week day, and time from epoch (all treated as numeric). This suffices to replicate the results I was seeing with time series telemetry data. This data consists of 6070 observations, with the following structure:

```
> head(burg)
# A tibble: 6 x 5
   floordate     N month  wday epoch
      <dttm> <int> <dbl> <dbl> <dbl>
1 2001-01-01    66     1     2     0
2 2001-01-02    68     1     3     1
3 2001-01-03    57     1     4     2
4 2001-01-04    55     1     5     3
5 2001-01-05    55     1     6     4
6 2001-01-06    64     1     7     5
```

To run this code, only a few libraries are required:

```
library(tidyverse)
library(xgboost)
library(lubridate)
```
First, I had to read in the [data file](https://github.com/dmlc/xgboost/files/1640976/burg_data.txt)
 (sorry, Github will only allow .txt files, not .csv) and generate data in a format that is `xgboost` friendly. I stored training and test data in both its original data frame format and cast as `xgb.DMatrix` in a single list called `xgb_data_list`. 

```
burg <- read.delim("burg_data.txt") %>%
   mutate(floordate = lubridate::mdy(floordate))

# raw train and test data
train_dat <- burg %>%
  filter(floordate < "2017-07-25") %>%
  select(-floordate)

test_dat <- burg %>%
  filter(floordate >= "2017-07-25") %>%
  select(-floordate)

# xgboost train and test matrices
train_sparse <- Matrix::sparse.model.matrix(N~.-1, data = train_dat)
train_label <- train_dat %>% pull(N) %>% as.numeric
train_matrix <- xgboost::xgb.DMatrix(data = train_sparse, label = train_label)

test_sparse <- Matrix::sparse.model.matrix(N~.-1, data = test_dat)
test_label <- test_dat %>% pull(N) %>% as.numeric
test_matrix <- xgboost::xgb.DMatrix(data = test_sparse, label = test_label)

xgb_data_list <- list(train_dat = train_dat, train_matrix = train_matrix, 
                      test_dat = test_dat, test_matrix = test_matrix)

```

Then I wrote a few wrapper functions around the xgboost model generations for ease of use when iterating over different values of the two hyperparameters and evaluating the model predictions on test data. 

```
fit_xgboost_train_model <- function(xgb_data_list,
                                    subsample = 0.8, colsample_bytree = 0.8){
  
  xgb_params = list(objective = "reg:linear",
                    eval_metric = "rmse",
                    booster = "gbtree",
                    eta = 0.15, 
                    max.depth = 3,
                    min_child_weight = 10,
                    subsample = subsample,
                    colsample_bytree = colsample_bytree,
                    nthread = 4)
  
  xgboost::xgb.train(params = xgb_params,
                     data = xgb_data_list$train_matrix, 
                     nrounds = 200) 
}


test_pred <- function(xgb_data_list, mod){
  
  df <- data_frame(pred = predict(mod, newdata = xgb_data_list$test_matrix))
  
  df_out <- df %>%
    bind_cols(xgb_data_list$test_dat)
  
  df_out
  
}

get_mod_data <- function(xgb_data_list, subsample, colsample_bytree){
  set.seed(seed = 100)
  mod_train <- fit_xgboost_train_model(xgb_data_list, 
                                       subsample = subsample, colsample_bytree = colsample_bytree)
  
  mod_imp <- xgboost::xgb.importance(feature_names = colnames(xgb_data_list$train_matrix), 
                                     model = mod_train) %>%
    select(Feature, Gain) 
  
  pred_df <- test_pred(xgb_data_list, mod = mod_train)
  
  data_frame(raw = length(mod_train$raw), imp_length = nrow(mod_imp), 
             MAE = mean(abs(pred_df$pred - pred_df$N)),
             subsample = subsample, colsample_bytree = colsample_bytree)
}
```

Finally, I utilized the `xgb_data_list` object and the functions above to iterate over several values of `subsample` and `colsample_bytree` (limited to 0.4 to 1.0 because < 0.333 not allowed for `colsample_bytree` with only three features).

I created a data frame of hyperparameter value combinations to iterate through. 

```
iter_df <- data_frame(subsample = c(rep(1, 7), seq(.9, .4, by = -.1)), 
                      colsample_bytree = c(seq(1, .4, by = -.1), rep(1, 6))) 

> iter_df
# A tibble: 13 x 2
   subsample colsample_bytree
       <dbl>            <dbl>
 1       1.0              1.0
 2       1.0              0.9
 3       1.0              0.8
 4       1.0              0.7
 5       1.0              0.6
 6       1.0              0.5
 7       1.0              0.4
 8       0.9              1.0
 9       0.8              1.0
10       0.7              1.0
11       0.6              1.0
12       0.5              1.0
13       0.4              1.0
```

Then for each combination, I stored information on the length of the training data raw vector, the mean absolute error (MAE) of each model, and the number of features included in the model importance data frame. 

```
run_test_df <- purrr::map_df(1:nrow(iter_df), function(row){
  subsample <- as.numeric(iter_df[row, "subsample"])
  colsample_bytree <- as.numeric(iter_df[row, "colsample_bytree"])
  
  get_mod_data(xgb_data_list = xgb_data_list, subsample = subsample, colsample_bytree = colsample_bytree)
})

saveRDS(run_test_df, "burg_mod_data_local.RDS")
```

This same code was ran locally and on the Linux VM. Results were combined and plotted, revealing the discrepancies I have described. 

```
local_dat <- readRDS("sched_holds/burg_mod_data_local.RDS") %>% mutate(OS = "OS X")
vm_dat <- readRDS("sched_holds/burg_mod_data_vm.RDS") %>% mutate(OS = "Linux")

run_tests <- bind_rows(local_dat, vm_dat) %>%
  gather(measure, value, -OS, -subsample, -colsample_bytree) %>%
  mutate(measure = case_when(measure == "raw" ~ "length of \n xgboost raw data",
                             measure == "MAE" ~ "mean absolute error",
                             measure == "imp_length" ~ "# features in \n importance matrix"))
```

**Varying subsample**

```
subs <- run_tests %>% filter(colsample_bytree == 1)

ggplot(subs, aes(x = subsample, y = value, col = OS, linetype = OS)) +
  geom_line() +
  facet_wrap(~measure, scales = "free_y") +
  ggtitle("Impact of varying subsample with subcolumn_bytree = 1") +
  scale_x_continuous(breaks = seq(0, 1, by = .2))
```
![vary_subsample](https://user-images.githubusercontent.com/10351038/35072474-fbed70b0-fba1-11e7-8542-04e2a6a09c2e.png)

**Vary colsample_bytree**

```
subc <- run_tests %>% filter(subsample == 1)

ggplot(subc, aes(x = colsample_bytree, y = value, col = OS, linetype = OS)) +
  geom_line() +
  facet_wrap(~measure, scales = "free_y") +
  ggtitle("Impact of varying subcolumn_bytree with subsample = 1") +
  scale_x_continuous("subcolumn_bytree", breaks = seq(0, 1, by = .2))
```
![vary_colsample](https://user-images.githubusercontent.com/10351038/35072473-fbdc3b6a-fba1-11e7-9e73-f3392f4d9f6d.png)

Source data: 
[burg_data.txt](https://github.com/dmlc/xgboost/files/1640976/burg_data.txt)

**Comment by xingyuezhiji** [comment]
agree！I have faced the same situation

**Comment by xingyuezhiji** [comment]
Agree！I have faced the same problem

**Comment by abelcastilloavant** [comment]
I've encountered the same problem.

**Comment by lorenzwalthert** [comment]
I seem to have the same issue. Why was this issue closed again?
